### DISCUSSION TITLE
how to get time step duration in a custom material object

### DISCUSSION THREAD

**Discussion by RWTHLHK** [idaholab/moose#27117] [Q&A Modules: Solid mechanics] [answered]
I need to develop a custom material that evolves with time. So I need to get the current timestep duration,  how can I do that in a custom material object?

**Comment by RWTHLHK** [comment, marked as answer]
i find it is _dt

**Reply by Traiwit** [reply]
how about 'current_time' ? i'm looking for it as well

Cheers
Trai

**Reply by GiudGiud** [reply]
if it's not `_time` then retrieve a reference to it from the problem in the constructor with `virtual Real & time() const { return _time; }`

**Reply by Traiwit** [reply]
got it @GiudGiud  thank you!

it was very difficult to find/search on the main website, I was making my own version of Exodus.C and I had to go to Output.C to look for all these functions

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
> it was very difficult to find/search on the main website

did you try the doxygen? it s better for code
https://mooseframework.inl.gov/docs/doxygen/moose/classes.html

**Reply by Mrxiongjy** [reply]
@Traiwit Dear Doctor, how did you finally get the current time in the source code, I also encountered this problem can not be solved

**Reply by GiudGiud** [reply]
The name of the problem attribute depends on the class you are using. But if it is called `_fe_problem`, and it's saved as a pointer,
then
```
_fe_problem->time()
```

**Reply by Mrxiongjy** [reply]
Thank you for your reply, this is my source code, I want to achieve a thermal expansion material core source code to get the current time, the following is my source code，Do you think it's okay to get time this way? I've tried many times，auto ti = _fe_problem.time();
```
#include "Relocationstrain.h"

registerMooseObject("pandasApp", Relocationstrain);
registerMooseObject("pandasApp", ADRelocationstrain);

template <bool is_ad>
InputParameters
RelocationstrainTempl<is_ad>::validParams()
{
  InputParameters params = ComputeThermalExpansionEigenstrainBaseTempl<is_ad>::validParams();
  params.addClassDescription("Computes eigenstrain due to thermal expansion "
                             "with a constant coefficient");
  return params;
}

template <bool is_ad>
RelocationstrainTempl<is_ad>::RelocationstrainTempl(const InputParameters & parameters)
  : ComputeThermalExpansionEigenstrainBaseTempl<is_ad>(parameters)
{
}

template <bool is_ad>
ValueAndDerivative<is_ad>
RelocationstrainTempl<is_ad>::computeThermalStrain()
{
  auto ti = _fe_problem.time();
  ValueAndDerivative<is_ad> y;
  Real Q;
  Real bu = 1.25 * ti * ti * 270 /
                (3.1416 * 10431 * 235 * (4.1 * std::pow(10, -3)) * (4.1 * std::pow(10, -3))) +
            0.001;
```

**Reply by GiudGiud** [reply]
does this compile?
I see _fe_problem in Material.C and MaterialBase.C which seems to be a base class of this

you might have to do `this->_fe_problem.time()` because of the templating, not sure

**Reply by Mrxiongjy** [reply]
This is what happens when I just compile
'_fe_problem' was not declared in this scope
   26 |   auto ti = _fe_problem->time();

**Reply by Mrxiongjy** [reply]
Thank you very, very much@GiudGiud
I made it

**Reply by Mrxiongjy** [reply]
> does this compile? I see _fe_problem in Material.C and MaterialBase.C which seems to be a base class of this
> 
> you might have to do `this->_fe_problem.time()` because of the templating, not sure

I changed it to this and it worked. Thank you
```
template <bool is_ad>
ValueAndDerivative<is_ad>
RelocationstrainTempl<is_ad>::computeThermalStrain()
{
  auto ti = this->_fe_problem.time();
  ValueAndDerivative<is_ad> y;
  Real Q;
  Real bu = 1.25 * ti * ti * 270 /
                (3.1416 * 10431 * 235 * (4.1 * std::pow(10, -3)) * (4.1 * std::pow(10, -3))) +
            0.001;
```
